ddev/ddev
# feat: add support for `n install auto`, add `N_PREFIX` cache, replace node repo with `n`, fixes #6418

## The Issue

- #6418

## How This PR Solves The Issue

From https://www.npmjs.com/package/n#specifying-nodejs-versions:

There is an `auto` label to read the target version from a file in the current directory, or any parent directory. `n` looks for in order:

- `.n-node-version`: version on single line. Custom to `n`.
- `.node-version`: version on single line. Used by multiple tools: [node-version-usage](https://github.com/shadowspawn/node-version-usage)
- `.nvmrc`: version on single line. Used by `nvm`.
- if no version file found, look for `engine` as below.

The `engine` label looks for a `package.json` file and reads the `engines` field to determine compatible Node.js. Requires an installed version of `node`, and uses `npx semver` to resolve complex ranges.

---

This PR adds:
- cache for `n` just like it works for `nvm`.
- full compatibility for `n install`, as previously it was run from a `Dockerfile` with no mounted code, which limited its use.

## Manual Testing Instructions

Review the docs https://ddev--6420.org.readthedocs.build/en/6420/users/configuration/config/#nodejs_version

```
ddev config --nodejs-version=auto
echo "10" > .nvmrc
ddev restart
ddev logs -s web # to see how it was installed
ddev exec node --version
```

Repeat the same with different options from https://www.npmjs.com/package/n#specifying-nodejs-versions

```
ddev config --nodejs-version=engine
echo '{
    "engines": {
        "node": "14.20"
    }
}' > package.json
ddev restart
ddev logs -s web # to see how it was installed
ddev exec node --version
```

## Automated Testing Overview



## Release/Deployment Notes
